File: src/neural_network/src/scripts/image_suscriber.py
#!/usr/bin/env python

import rospy
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError
import cv2

def main_process(image):
    gray_image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    equalized_image = cv2.equalizelist(gray_image)
    cv2.imshow("Equalized image P1", equalized_image)
    cv2.waitKey(1)

def imageCallback(msg):
    try:
        bridge = CvBridge()
        cv_image = bridge.imgmsg_to_cv2(msg,desired_encoding="bgr8")
        cv2.imshow("Image Subscriber PY",cv_image)
        cv2.waitKey(1)
        main_process(cv_image)
    except CvBridgeError as e:
        rospy.logerr("CvBridge Error: %s",e)

if __name__ == "__main__":
    rospy.init_node("image_subscriber_node_py")
    rospy.Subscriber("/camera/image_raw", Image, imageCallback)
    rospy.spin()

File: src/neural_network/src/scripts/image_suscriber.cpp
#include <ros/ros.h>
#include <opencv2/highgui/highgui.hpp>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/Image.h>

void main_process(cv::Mat& image){
    cv::Mat gray_image;
    cv::cvtColor(image, gray_image, cv::COLOR_BGR2GRAY);
    cv::equalizeHist(gray_image, gray_image);
    cv::imshow("Equalize image CPP", gray_image);
    cv::waitKey(1);
}


void imageCallback(const sensor_msgs::ImageConstPtr msg){
    try {
        cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
        cv::imshow("Image Suscriber CPP", cv_ptr->image);
        cv::waitKey(1);
        main_process(cv_ptr->image);
    } catch (cv_bridge::Exception& e){
        ROS_ERROR("cv_bridge exception: %s", e.what());
    }
}

int main(int argc, char** argv){
    ros::init(argc, argv, "image_suscribe_node_cpp");
    ros::NodeHandle nh;

    ros::Subscriber sub = nh.subscribe("camera/image_raw",1,imageCallback);
    ros::spin();
    return 0;

}